src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -161,7 +162,11 @@ internal bool TryGetPolymorphicTypeInfoForRootType(object rootValue, [NotNullWhe
             Type runtimeType = rootValue.GetType();
             if (runtimeType != JsonTypeInfo.ObjectType)
             {
+                // To determine the contract for an object value:
+                // 1. Find the JsonTypeInfo for the runtime type with fallback to the nearest ancestor, if not available.
+                // 2. If the resolved type is deriving from a polymorphic type, use the contract of the polymorphic type instead.
                 polymorphicTypeInfo = GetTypeInfoForRootType(runtimeType, fallBackToNearestAncestorType: true);
+                polymorphicTypeInfo = polymorphicTypeInfo.AncestorPolymorphicType ?? polymorphicTypeInfo;
```

Comment:
This double set of the variable looked weird to me. and it took me a second to figure out what it was doing.

I wonder if written this way would be more readable/straight-forward.

```suggestion
                if (polymorphicTypeInfo.AncestorPolymorphicType is JsonTypeInfo ancestorPolymorphicType)
                {
                    polymorphicTypeInfo = ancestorPolymorphicType;
                }
```